runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/getdata/ApplianceGetDataClient.java

```diff
@@ -0,0 +1,253 @@
+    this.pendingReadBatches = new ArrayList<>();
+    this.activeReadThreads = 0;
+  }
+
+  public static GetDataClient create(
+      ApplianceWindmillClient windmillClient, ThrottlingGetDataMetricTracker getDataMetricTracker) {
+    return new ApplianceGetDataClient(windmillClient, getDataMetricTracker);
+  }
+
+  @Override
+  public Windmill.KeyedGetDataResponse getStateData(
+      String computation, Windmill.KeyedGetDataRequest request) {
+    try (AutoCloseable ignored =
+        getDataMetricTracker.trackSingleCallWithThrottling(
+            ThrottlingGetDataMetricTracker.Type.STATE)) {
+      SettableFuture<Windmill.KeyedGetDataResponse> response = SettableFuture.create();
+      ReadBatch batch = addToReadBatch(new QueueEntry(computation, request, response));
+      if (batch != null) {
+        issueReadBatch(batch);
+      }
+      return response.get();
+    } catch (Exception e) {
+      throw new GetDataException(
+          "Error occurred fetching state for computation="
+              + computation
+              + ", key="
+              + request.getShardingKey(),
+          e);
+    }
+  }
+
+  @Override
+  public Windmill.GlobalData getSideInputData(Windmill.GlobalDataRequest request) {
+    try (AutoCloseable ignored =
+        getDataMetricTracker.trackSingleCallWithThrottling(
+            ThrottlingGetDataMetricTracker.Type.STATE)) {
+      return windmillClient
+          .getData(Windmill.GetDataRequest.newBuilder().addGlobalDataFetchRequests(request).build())
+          .getGlobalData(0);
+    } catch (Exception e) {
+      throw new GetDataException(
+          "Error occurred fetching side input for tag=" + request.getDataId(), e);
+    }
+  }
+
+  /**
+   * Appliance sends heartbeats (used to refresh active work) as KeyedGetDataRequests. So we must
+   * translate the HeartbeatRequest to a KeyedGetDataRequest.
+   */
+  @Override
+  public void refreshActiveWork(Map<HeartbeatSender, Heartbeats> heartbeats) {
+    Map.Entry<HeartbeatSender, Heartbeats> heartbeat =
+        Iterables.getOnlyElement(heartbeats.entrySet());
+    HeartbeatSender heartbeatSender = heartbeat.getKey();
+    Heartbeats heartbeatToSend = heartbeat.getValue();
+
+    if (heartbeatToSend.heartbeatRequests().isEmpty()) {
+      return;
+    }
+
+    try (AutoCloseable ignored = getDataMetricTracker.trackHeartbeats(heartbeatToSend.size())) {
+      heartbeatSender.sendHeartbeats(heartbeatToSend);
+    } catch (Exception e) {
+      throw new GetDataException("Error occurred refreshing heartbeats=" + heartbeatToSend, e);
+    }
+  }
+
+  @Override
+  public synchronized void printHtml(PrintWriter writer) {
+    getDataMetricTracker.printHtml(writer);
+    writer.println("  Read threads: " + activeReadThreads);
+    writer.println("  Pending read batches: " + pendingReadBatches.size());
+  }
+
+  private void issueReadBatch(ReadBatch batch) {
+    try {
+      Preconditions.checkState(batch.startRead.get());
```

Comment:
// Possibly block until the batch is allowed to start.

Reply:
done